File: src/com/isa/tree/_offerII_044_largestValues.java
package com.isa.tree;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class _offerII_044_largestValues {

    public List<Integer> largestValues(TreeNode root) {
        List<Integer> list = new ArrayList();
        if (root == null) return list;

        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            int max = Integer.MIN_VALUE;
            int levelSize = queue.size();
            for (int i = 0; i < levelSize; i++) {
                TreeNode node = queue.poll();
                if (node.val > max) max = node.val;

                if (node.left != null) queue.offer(node.left);
                if (node.right != null) queue.offer(node.right);
            }
            list.add(max);
        }

        return list;
    }
}

File: src/com/isa/tree/_offerII_044_largestValues.kt
package com.isa.tree

import java.util.LinkedList
import java.util.Queue

fun largestValues(root: TreeNode?): List<Int> {
    val list: MutableList<Int> = mutableListOf()

    if (root == null) return list
    val queue: Queue<TreeNode> = LinkedList()
    queue.offer(root)

    while (!queue.isEmpty()) {
        var max = Int.MIN_VALUE
        val levelSize = queue.size

        for (i in 0 until levelSize) {
            val node = queue.poll()
            if (node.`val` > max) max = node.`val`

            node.left?.let {
                queue.offer(it)
            }

            node.right?.let {
                queue.offer(it)
            }
        }
        list.add(max)
    }
    return list
}